tests/transform/test_read_fragments.py

```diff
@@ -0,0 +1,71 @@
+from datetime import date
+from dataclasses import dataclass
+
+from google.cloud import bigquery
+from google.api_core.exceptions import BadRequest
+
+import apache_beam as beam
+
+from apache_beam.testing.test_pipeline import TestPipeline
+
+from pipe_segment.transform.read_fragments import ReadFragments
+
+
+class RowMock:
+    min_suffix = "20240301"
+
+
+class ResultMock:
+    def result(self):
+        return [RowMock()]
+
+
+class ResultErrorMock:
+    def result(self):
+        raise BadRequest("Bad request")
+
+
+@dataclass
+class BigQueryClientMock:
+    project: str
+
+    def query(self, query):
+        if self.project is None:
+            return ResultErrorMock()
+
+        return ResultMock()
+
+
+class ReadFromBigQueryMock(beam.io.ReadFromBigQuery):
+    def expand(self, pcoll):
+        return pcoll | beam.Create([])
+
+
+def test_read_fragments(monkeypatch):
+    # TODO: replace this monkey patch when design allows for more easy testing.
+    monkeypatch.setattr(bigquery, "Client", BigQueryClientMock)
+    monkeypatch.setattr(beam.io, "ReadFromBigQuery", ReadFromBigQueryMock)
+
+    dummy_source = ""
+
+    # Test raising BadRequest not create if missing
+    op = ReadFragments(dummy_source, date.today(), date.today())
+    with TestPipeline() as p:
+        p | op
+
+    # Test raising BadRequest and create if missing
+    op = ReadFragments(dummy_source, date.today(), date.today(), create_if_missing=True)
+    with TestPipeline() as p:
+        p | op
+
+    # Test without raising BadRequest
+    op = ReadFragments(dummy_source, date.today(), date.today(), project="dummy")
+    with TestPipeline() as p:
+        p | op
+
+    # Test without start date
+    op = ReadFragments(dummy_source, None, date.today(), project="dummy")
+    with TestPipeline() as p:
+        p | op
+
+    # TODO: add asserts about outputs
```

Comment:
I don't like these kinds of IO tests. I agree with them right now though, because we are doing a bunch of things inside the read fragments transform. But ideally, IO should only have IO, and we should be using standard `BigQueryRead` transforms that only read data out from BigQuery, and do whatever we need to do with that data in more testeable later transforms. WDYT?

Reply:
Even if you don't have a wrapper around a built-in IO transform, you will need to test your code that uses that transform, so I don't know if I understand what you mean with this kind of IO tests. If you don't agree with the existence of this particular transform, I think that's another discussion. Regarding that, I don't think it's forbidden to have a wrapper around a built-in IO transform. Sometimes you do need to extend and add functionality.